pkg/kubelet/cm/cpumanager/topology_hints.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cpumanager
+
+import (
+	"k8s.io/api/core/v1"
+	"k8s.io/klog"
+	"math"
+
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpumanager/topology"
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/socketmask"
+)
+
+func (m *manager) GetTopologyHints(pod v1.Pod, container v1.Container) []topologymanager.TopologyHint {
+	var cpuHints []topologymanager.TopologyHint
+	for resourceObj, amountObj := range container.Resources.Requests {
+		resource := string(resourceObj)
+		if resource != "cpu" {
+			continue
+		}
+
+		// Get a count of how many CPUs have been requested
+		requested := int(amountObj.Value())
+		klog.Infof("[cpumanager] Guaranteed CPUs detected: %v", requested)
+
+		// Discover topology in order to establish the number
+		// of available CPUs per socket.
+		topo, err := topology.Discover(m.machineInfo)
+		if err != nil {
+			klog.Infof("[cpu manager] error discovering topology")
+			continue
+		}
+		reserved := m.getReservedCPUs(topo)
+		containerID, _ := findContainerIDByName(&pod.Status, container.Name)
+		availableCPUs := m.getAvailableCPUs(containerID, reserved)
+		cpuAccum := newCPUAccumulator(topo, availableCPUs, requested)
+		socketCount := topo.NumSockets
+		cpusPerSocket := getCPUsPerSocket(socketCount, cpuAccum)
+		klog.Infof("[cpumanager] Number of Available CPUs per Socket: %v", cpusPerSocket)
+
+		// Build cpuHintsAffinity []TopologyHint from all possible socket combinations where the
+		// request for CPUs can be granted, agnostic to what hints are/are not Preferred.
+		cpuHintsAffinity := getCPUHintsAffinity(cpusPerSocket, requested)
+
+		if cpuHintsAffinity != nil {
+			// Create new CPU Accumulator in order to find mostCPUs.
+			cpuAccumStandard := newCPUAccumulator(topo, topo.CPUDetails.CPUs().Difference(reserved), requested)
+			// mostCPUs is the largest number of CPUs on any one socket assuming
+			// all CPUs (minus reserved CPUs) are available.
+			mostCPUs := getMostCPUs(cpuAccumStandard, socketCount)
+
+			// Assign 'Preferred: true/false' values for each hint.
+			cpuHints = getCPUHintsPreferred(cpuHintsAffinity, requested, mostCPUs)
+		}
+	}
+	klog.Infof("[cpumanager] Topology Hints for pod: %v", cpuHints)
```

Comment:
In the logs I'm getting this:

2019-06-21T21:51:45.041 controller-1 kubelet[2159067]: info I0621 21:51:45.041010 2159067 topology_hints.go:106] [cpumanager] Topology Affinities for pod: [{1000000000000000000000000000000000000000000000000000000000000000 true} {0100000000000000000000000000000000000000000000000000000000000000 true} {1100000000000000000000000000000000000000000000000000000000000000 true}]

First, I think those bitmasks should be the other way around with NUMA node 1 being the rightmost bit instead of leftmost.  Second, I think we should not print the high-order zero bits to minimize what gets printed, since the common cases will likely be 1/2/4 NUMA nodes.

Reply:
Reversing the bit order is addressed in https://github.com/kubernetes/kubernetes/pull/78587.

Regarding stripping the high order bits. I would probably not strip them completely. Instead, I would strip them down to lengths of 2, 4, 8, and then multiples of 8 up to 64 depending on which bits are set.